ecommerce/factories.py

```diff
@@ -87,3 +89,41 @@ class CoursePriceFactory(DjangoModelFactory):
 
     class Meta:  # pylint: disable=missing-docstring,no-init,too-few-public-methods,old-style-class
         model = CoursePrice
+
+
+class CouponFactory(DjangoModelFactory):
+    """Factory for Coupon"""
+    amount_type = FuzzyChoice(Coupon.AMOUNT_TYPES)
+    coupon_code = FuzzyChoice([None, FuzzyText().fuzz()])
+    num_redemptions_per_user = FuzzyInteger(1, 10)
+    num_coupons_available = FuzzyInteger(1, 100)
+
+    class Meta:  # pylint: disable=missing-docstring,no-init,too-few-public-methods,old-style-class
+        model = Coupon
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
```

Comment:
I think you're trying to be too clever, here. Factory Boy is best used with reasonable, static values that can be easily extended and overwritten when the factory is used in a test. Trying to override the `_create()` method usually leads to awkward interactions down the line, when trying to use this factory with others. Here's what I'd suggest instead:

```python
import factory
from factory.django import DjangoModelFactory


class CouponFactory(DjangoModelFactory):
    coupon_code = FuzzyText()
    amount_type = 'percent-discount'
    amount = 0.5
    num_coupons_available = 15
    num_redemptions_per_user = 2

    class Meta:
        model = Coupon

    content_object = factory.SubFactory(ProgramFactory)

    @factory.post_generation
    def content_type(self, create, extracted, **kwargs):
        return self.content_object.__class__

    @factory.post_generation
    def object_id(self, create, extracted, **kwargs):
        return self.content_object.id

    class Params:
        percent = factory.Trait(
            amount_type='percent-discount',
            amount=0.5, # 50% off
        )
        fixed = factory.Trait(
            amount_type='fixed-discount',
            amount=100, # $100 off
        )
        program = factory.Trait(
            content_object=factory.SubFactory(ProgramFactory)
        )
        course = factory.Trait(
            content_object=factory.SubFactory(CourseFactory)
        )
        course_run = factory.Trait(
            content_object=factory.SubFactory(CourseRunFactory)
        )
```

This defines some reasonable, static defaults while easily allowing the caller to override them using the [trait](https://factoryboy.readthedocs.io/en/latest/reference.html#traits) system. For example, to make a coupon for a course instead of a program, just run:
```python
coupon = CouponFactory(course=True)
```
Or, if you already have a course object that you want to generate a coupon for:
```python
coupon = CouponFactory(content_object=my_course)
```

By default, these coupons are for 50% off. To adjust the percentage, just change it in the caller:
```python
coupon = CouponFactory(amount=0.3)  # 30% off
```
Or to change it to a fixed amount, use the `fixed` trait:
```python
coupon = CouponFactory(fixed=True, amount=150)  # $150 off
```
You can even use the `percent` trait for clarity:
```python
coupon = CouponFactory(percent=True, amount=0.8)  # 80% off
```

And of course, all the other properties can be easily overridden as well, as per usual for Factory Boy factories:
```python
coupon1 = CouponFactory(coupon_code="mr.r0bot")
coupon2 = CouponFactory(num_coupons_available=9000, num_redemptions_per_user=10)
coupon3 = CouponFactory(activation_date=datetime(2018, 1, 1, 0, 0, 0), expiration_date=datetime(2019, 1, 1, 0, 0, 0))
coupon4 = CouponFactory(enabled=False)
```